### DISCUSSION TITLE
Is it forced to using Await to make React.Suspense's fallback work?

### DISCUSSION THREAD

**Discussion by Toshinaki** [remix-run/react-router#10697] [Q&A] [answered]
I'm experimenting with new v6 features.
After added some delay to my mock apis, I noticed the page went blank until the `loader` returns data.

I did some searching and reviewed the docs again. It seems that the only way to make Suspense work is to use `Await`?
Is that true?

I'm not an expert at js or React, but this seems not very good, IMHO.
I could simply wrap all components with Suspense, and write no code that's not related to the pages' logic. 
But now I have to add `Await` to all the components.

I'm also using react-query to manage server state. Didn't find any docs about combine react-query & loader & Await.

**Comment by brophdawg11** [comment, marked as answer]
For the initial execution of loaders in the browser, you can show a loading indicator via [`fallbackElement`](https://reactrouter.com/en/main/routers/router-provider#fallbackelement).  The `Await`/`Suspense` combination is for [Deferred Data](https://reactrouter.com/en/main/guides/deferred).

**Reply by Toshinaki** [reply]
yes, I've noticed fallbackElement during searching through docs and google.
though I want to wrap for example a sidebar (as a child path of the main path) of a list page with suspense, 
so when click some item in a list, the sidebar opens immediately, and if data is not ready, Suspense fallback displays until data is ready.
and fallback shows when switch between list items, if the item data is not ready.
A fullpage fallback is just too rough.

**Reply by brophdawg11** [reply]
Yep - that's what `defer`/`Await`/`Suspense` is for.  You can `defer` the load of the individual item and the navigation will complete immediately and trigger the `Suspense` around your `Await`.

```jsx
function itemLoader() {
  return defer({
    item: getItem() // not awaited
  });
}

function Sidebar() {
  let data = useLoaderData();
  return (
    <React.Suspense fallback={<p>Loading...</p>}>
      <Await resolve={data.item}>
        {item => <p>{item}</p>}
      </Await>
    </React.Suspense>
  );
}

**Reply by MorningLit** [reply]
Hi @brophdawg11 , I'm trying to get React suspense to work with react router dom and I'm not sure how?

If I want to do some data processing after receiving the data, how do I go about doing it? I tried doing a useMemo inside the Await component, but it doesn't seem to work, eg.

```typescript
function Sidebar() {
  let data = useLoaderData();
  return (
    <React.Suspense fallback={<p>Loading...</p>}>
      <Await resolve={data.item}>
        <Component/>
      </Await>
    </React.Suspense>
  );
}
function Component() {
    let item = useAsyncValue();
    const processed = useMemo(async ()=>{
        processItem(await item);
    },[item]);
    // how?
    return <p>{processed}</p>;
}
```

**Reply by brophdawg11** [reply]
That should be fine - can you provide a reproduction of what exactly isn't working?

**Reply by MorningLit** [reply]
Oh, maybe passing the processed data to outlet doesn't work?
After processing I pass the data to the Outlet component and when my other child routes tried to read from the useOutletContext(), it crashed my app, my data in Outlets are now promises

**Reply by brophdawg11** [reply]
It's hard to say without a minimal reproduction of the issue.

**Reply by MorningLit** [reply]
Made a code sandbox, https://codesandbox.io/s/trusting-dhawan-rfmwql?file=/src/Home.js , please let me know if I made any mistakes 🥸

**Reply by brophdawg11** [reply]
ah, ok, you are getting a promise because you're creating one by using `async`/`await` in your `useMemo` - if you give an async function to `useMemo` it will return you a `Promise`.  However, `data` is already resolved there since it's a child of `Await` so you can just operate on the data directly: https://codesandbox.io/s/youthful-cookies-fsg8c4

However, I think a better approach would be to do this data logic in the loader and keep it out of the UI layer: https://codesandbox.io/s/practical-dewdney-7m6dgx?file=/src/index.js

**Reply by MorningLit** [reply]
Thanks for the help! 
One thing I would like to ask is that how do I get the fallback component to load? My main use case of using the Suspense with react-router-dom is to display a loading component when it is fetching the data, but it is not showing 😢

**Reply by brophdawg11** [reply]
The usage of defer is outlined in the [docs](https://reactrouter.com/en/main/guides/deferred)
